### ISSUE TITLE
filtergroupslist not working with NTLM authentication

### ISSUE THREAD

**Issue by ai99** [e2guardian/e2guardian#130]
When using LDAP authentication everything works fine.
When using NTLM authentication filtergroups does not work.

Create a e2guardian with 2 filtergroups.
blacklist youtube.com on filter1, surf on youtube with <user>, youtube is forbidden, Ok.
assign filter2 to <user>, surf on youtube, access is still denied.

Note that access denied page returned by e2guardian is truncated.

![capture d ecran_2016-06-06_11-09-48](https://cloud.githubusercontent.com/assets/15774019/15818091/95dae70e-2bdb-11e6-9fed-22bafa1bd476.png)
![capture d ecran_2016-06-06_11-10-47](https://cloud.githubusercontent.com/assets/15774019/15818090/95d90ca4-2bdb-11e6-919b-1c0eb73abb81.png)

**Comment by fredbcode** [comment]
NTLM or kerberos ?
Can you see the user account with tcpdump ?

**Comment by ai99** [comment]
NTLM auth, squid.conf : 
`auth_param ntlm program /usr/lib/squid3/ntlm_smb_lm_auth -a  dompedago/nt4pdc`

With tshark :
` 72  36.521067    10.1.2.50 -> 10.1.2.1     HTTP 672 GET http://www.youtube.com/ HTTP/1.1 , NTLMSSP_AUTH, User: DOMPEDAGO\ADMIN`

**Comment by fredbcode** [comment]
You see DOMPEDAGO in logs (e2guardian's log)?

**Comment by ai99** [comment]
No, I only see the username :
`2016-06-06T14:48:07.657573+02:00 amon.etb1.lan e2guardian0[778]: "2016.6.6 14:48:07","admin","10.1.2.50","http://etb1.ac-test.fr/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom","","GET","0","0","","5","304","-","","","","","-","-",""`

For information, on one request, squid log makes 3 lines :
```
2016-06-06T14:50:19.169801+02:00 amon.etb1.lan squid3: 1465217419.169      0 10.1.2.50 TCP_DENIED/407 4760 GET http://etb1.ac-test.fr/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom - HIER_NONE/- text/html
2016-06-06T14:50:19.183933+02:00 amon.etb1.lan squid3: 1465217419.183      5 10.1.2.50 TCP_DENIED/407 4927 GET http://etb1.ac-test.fr/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom - HIER_NONE/- text/html
2016-06-06T14:50:19.203806+02:00 amon.etb1.lan squid3: 1465217419.203     14 10.1.2.50 TCP_REFRESH_UNMODIFIED/304 336 GET http://etb1.ac-test.fr/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom dompedago\\admin HIER_DIRECT/192.168.0.31 -

```

I suppose e2guardian manages the connection as an unknown user (thus puts the connection on filter1) due to NTLM handshake (squid/e2guardian knows username only at the 3rd line). 
Note that this bug exists also on Kerberos authentication.

**Comment by fredbcode** [comment]
> No, I only see the username : 
> 2016-06-06T14:48:07.657573+02:00 amon.etb1.lan e2guardian0[778]: "2016.6.6 14:48:07","admin","10.1.2.50","http://etb1.ac-test.fr> /guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site>%3a%20youtube%2ecom","","GET","0","0","","5","304","-","","","","","-","-","" 

Ho admin is the username ? if yes it just should in groups files

admin=filter2 -> or whatever group you are using 

14:48:07","admin","10.1.2.50","http://etb1.ac-test.fr show that the user is right detected by e2guardian

What do you have in your groups file ?

**Comment by ai99** [comment]
Yes, "admin" is the username.

```
# cat common/filtergroupslist
admin=filter5
```
filter1 = default (where youtube.com is blacklisted)
filter5 =  (where youtube.com is allowed/not blacklisted)
(just for info, actually we have 3 other filters : filter2=bypasslink, filter3=banned, filter4=blanket block)

**Comment by fredbcode** [comment]
Can you name the group in the log ?

in e2guardianf5.conf 
groupname="groupe5"

And check e2guardian.conf
filtergroups = 5
filtergroupslist = '/YOURPATH/common/filtergroupslist'

And see what happen in e2 log

**Comment by ai99** [comment]
After having named groups :
```
2016-06-08T10:44:24.404177+02:00 amon.etb1.lan e2guardian0[9739]: Reconfiguring E2guardian: gentle reload starting
2016-06-08T10:44:35.600538+02:00 amon.etb1.lan e2guardian0[9900]: Started sucessfully.
2016-06-08T10:44:45.956211+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","admin","10.1.2.50","http://www.youtube.com","","GET","0","0","","5","301","-","",""group5"","","","-","-",""
2016-06-08T10:44:45.960545+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","-","10.1.2.50","https://www.youtube.com:443","*Accès interdit* Blocked HTTPS site: youtube.com","CONNECT","0","0","","1","403","-","",""group1"","","","-","-",""
2016-06-08T10:44:45.969658+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","admin","10.1.2.50","http://192.168.0.31/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom","*SCANNED* ","GET","169","0","","5","302","-","",""group5"","","","-","-",""
2016-06-08T10:44:45.976950+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","admin","10.1.2.50","http://etb1.ac-test.fr/guardian.html?DENIEDURL==https%3a%2f%2fwww%2eyoutube%2ecom::IP==10.1.2.50::USER==-::CATEGORIES==::REASON==Blocked%20HTTPS%20site%3a%20youtube%2ecom","*SCANNED* ","GET","1868","0","","5","200","text/html","",""group5"","","","-","-",""
```

**Comment by fredbcode** [comment]
Ok right, the first request is good

admin = group5 so no problem, as you can see this is a **HTTP** request
2016-06-08T10:44:45.956211+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","admin","10.1.2.50","http://www.youtube.com","","GET","0","0","","5","301","-","",""group5"","","","-","-",""

The second is false, no name so no group, this is a **HTTPS** requests 

2016-06-08T10:44:45.960545+02:00 amon.etb1.lan e2guardian0[9901]: "2016.6.8 10:44:45","-","10.1.2.50","https://www.youtube.com:443","*Accès interdit* Blocked HTTPS site: youtube.com","CONNECT","0","0","","1","403","-","",""group1"","","","-","-",""

Your problem needs more investigations, why there is a difference between the first and the second request from your browser ?

As far I can tell this is not related with e2guardian, I'm pretty sure that with a tcpdump you show no user in your **HTTPS** request

**Comment by ai99** [comment]
If I can do something else to help debugging, or if you think this bug has to be reported somewhere else, please don't hesitate to let me know.
This is a "a solution must be found" case here.

**Comment by ai99** [comment]
I've tried different options, including using tshark instead, but no success.

How can you see username with tcpdump ?

**Comment by fredbcode** [comment]
tcpdump -i ethx -A -n port 3128 | grep "Proxy-Authorization:" 

ethx = an interface before e2guardian
3128 = e2's port

After that the NTLM must be decoded with base64, 
There is also a plugin (Firefox) name livehttpheaders it could help 

If you use squid directly do you see any difference ? And with another HTTPS website ?

**Comment by fredbcode** [comment]
> If you use squid directly do you see any difference ? 

The answer is above 

2016-06-06T14:54:33.752863+02:00 amon.etb1.lan squid3: 1465217673.752     67 10.1.2.50 TCP_MISS/301 855 GET http://www.youtube.com/ dompedago\\admin HIER_DIRECT/216.58.210.238 text/html
2016-06-06T14:54:33.758105+02:00 amon.etb1.lan squid3: 1465217673.758      0 10.1.2.50 TCP_DENIED/407 3768 CONNECT www.youtube.com:443 - HIER_NONE/- text/html

As you can see **"TCP_DENIED/407"** with youtube in SLL mode 

So I guess directly connected to squid the problem should still the same

**Comment by ai99** [comment]
Acutally, 407 are normal, they are part of the NTLM authentication process and happened every time like this :
2016-09-16T11:47:41.201010+02:00 amon.etb1.lan squid3: 1474019261.200      0 10.1.2.51 TCP_DENIED/407 3951 GET http://perdu.com/ - HIER_NONE/- text/html
2016-09-16T11:47:41.237406+02:00 amon.etb1.lan squid3: 1474019261.237      7 10.1.2.51 TCP_DENIED/407 4118 GET http://perdu.com/ - HIER_NONE/- text/html
2016-09-16T11:47:41.262740+02:00 amon.etb1.lan squid3: 1474019261.262     18 10.1.2.51 TCP_MEM_HIT/200 596 GET http://perdu.com/ dompedago\\admin HIER_NONE/- text/html

- When accessing a regular website, authentication works well (login is logged by e2guardian).
- When accessing a denied website,  authentication works well, login is showed on e2 denied page.
- When accessing a website that is denied in default filtergroup and login is in another filtergroup (where youtube is allowed) does not work, e2guardian returns an incomplete denied page.

**Comment by ai99** [comment]
Maybe this helps : 
- filtergroup1 : "youtube.com denied", filtergroup2 : "all access"
  - user=filter1 => e2guardian Denied page with all info (username, IP, URL, etc.)
  - user=filter2 => **e2guardian Denied page truncated _(while youtube.com should be accessible)_**

- filtergroup1 : "all access", filtergroup2 : "youtube.com denied"
  - user=filter1 => youtube.com is accessible
  - user=filter2 => e2guardian Denied page with all info (username, IP, URL, etc.)

**Comment by ai99** [comment]
I've tested Firefox livehttpheaders plugin, whatever the site I surf on, it never shows authentication informations.

If I use squid directly, it works, like it work in the 3 cases listed in previous comment.

There is no problem with http websites, only with httpS.

I'm trying to get more information with tcpdump and NTLM auth decoding.

**Comment by ai99** [comment]
I've made 2 captures with wireshark : 

- wireshark_http.pcapng that shows a user in filter2 connecting an HTTP website forbidden in filter1, user can correctly access the website.
- wireshark_https.pcapng that shows a user in filter2 connecting an HTTPS website forbidden in filter1 which actually leads to a truncated e2guardian denied page while user should be allowed to surf on the website.

[wireshark_http.pcapng.gz](https://github.com/e2guardian/e2guardian/files/482712/wireshark_http.pcapng.gz)
[wireshark_http.pcap.gz](https://github.com/e2guardian/e2guardian/files/482710/wireshark_http.pcap.gz)

[wireshark_https.pcapng.gz](https://github.com/e2guardian/e2guardian/files/482713/wireshark_https.pcapng.gz)
[wireshark_https.pcap.gz](https://github.com/e2guardian/e2guardian/files/482717/wireshark_https.pcap.gz)

**Comment by fredbcode** [comment]
Yes, there is something very weird

As you can see with the good request (HTTP) the user sends his identification to squid, line 23 

[truncated]Proxy-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAHUAAAAYABgAjQAAAAkACQBYAAAABQAFAGEAAAAPAA8AZgAAAAAAAAClAAAABoIAAgoAOTgAAAAPW5JlKFL7ss+/MbDyGm3TGERPTVBFREFHT0FETUlOREVTS1RPUC01U1BINjk1xhOzVXku/LE90zSMBqaNIcaygWhm3w7ExhOzVXku/LE90

But not at all with youtube (HTTPS), line 32
So, of course, e2guardian didn't know the user and move the request to filter 1 (default group), this is normal.

There is something more strange with your configuration 
This is not the same browser who speaks with E2guardian !

Line 16, youtube first request HTTP, answer 302 code -> redirect to HTTPS 
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:42.0) Gecko/20100101 Firefox/42.0\r\n

Firefox 42

Line 32, youtube second request HTTPS
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64; rv:43.0) Gecko/20100101 Firefox/43.0\r\n 

Two seconds later Firefox 43 ? -> Without identification field
How do You explain This ?

What happen when you try directly https://www.youtube.com and not http://www.youtube.com ?

**Comment by ai99** [comment]
Here are 3 new captures with the latest version of Firefox (sorry, I didn't check that in my previous test).
* http, like previously
* https, I accessed directly https://youtube.com (denied in filter1 but with a user in filter2), the truncate e2g denied page is the same, only the capture may contain different informations
* https_allowed, like http access to an allowed website but with https

[wireshark_http.pcapng.gz](https://github.com/e2guardian/e2guardian/files/484632/wireshark_http.pcapng.gz)
[wireshark_https.pcapng.gz](https://github.com/e2guardian/e2guardian/files/484633/wireshark_https.pcapng.gz)
[wireshark_https_allowed.pcapng.gz](https://github.com/e2guardian/e2guardian/files/484980/wireshark_https_allowed.pcapng.gz)

**Comment by fredbcode** [comment]
>    * https, I accessed directly https://youtube.com (denied in filter1 but with a user in filter2), the truncate e2g denied page is the same, only the capture may contain different informations 

Like before there is no

    HTTP/1.1 407 Proxy Authentication Required

Do you have some acl in squid.conf ? Can you make a wireshark directly connected to squid ?
It seems that there is no authentication needed for youtube (https)
 

>    * https_allowed, like http access to an allowed website but with https 

Yes this is correct

407 and FF returns the user account

**Comment by ai99** [comment]
Here a capture on squid(8080).
[wireshark_https_squid_8080.pcapng.gz](https://github.com/e2guardian/e2guardian/files/485162/wireshark_https_squid_8080.pcapng.gz)

**Comment by ai99** [comment]
And another made with google chrome.
[wireshark_https_squid_8080_ggle-chrome.pcapng.gz](https://github.com/e2guardian/e2guardian/files/485190/wireshark_https_squid_8080_ggle-chrome.pcapng.gz)

**Comment by fredbcode** [comment]
I can't open your first file
About the second your are stopped by squid ? There is no NTLM account pushed by chrome

**Comment by ai99** [comment]
This file should be OK
[wireshark_https_squid_8080.pcapng.gz](https://github.com/e2guardian/e2guardian/files/485234/wireshark_https_squid_8080.pcapng.gz)

With GoogleChrome I get : 
![capture d ecran_2016-09-21_16-12-57](https://cloud.githubusercontent.com/assets/15774019/18714549/686a92a8-8016-11e6-8641-7d0adba40d3e.png)

**Comment by fredbcode** [comment]
Your file contains some e2 messages ?
 
But we are enough informations with chrome, this is a squid problem, there is no identification with youtube (HTTPS) and E2guardian set the user to filter1, this is an attended behaviour.

**Comment by fredbcode** [comment]
I guess you have only 407 results for youtube (I'm speaking with e2 disabled)

2016-06-06T14:54:33.758105+02:00 amon.etb1.lan squid3: 1465217673.758 0 10.1.2.50 TCP_DENIED/407 3768 CONNECT www.youtube.com:443 - HIER_NONE/- text/html

No 200 ? No 304 ?

**Comment by fredbcode** [comment]
2016-09-16T11:47:41.262740+02:00 amon.etb1.lan squid3: 1474019261.262 18 10.1.2.51 **TCP_MEM_HIT/200** 596 GET http://perdu.com/ **dompedago\admin** HIER_NONE/- text/html

Without this value E2 doesn't knows what to do with your request

**Comment by ai99** [comment]
For me the problem is that e2 does not wait the NTLM authentication process to complete before it checks the username.

When YTB is allowed (or on every allowed https site), one can see multiple access to YTB before getting : 
`HTTP/1.1 200 Connection established`

When YTB is denied, e2 switches directly to filter1 denied page, just after : 
`HTTP/1.1 407 Proxy Authentication Required  (text/html)`

![capture d ecran_2016-09-21_16-43-54](https://cloud.githubusercontent.com/assets/15774019/18715983/6c8ab890-801b-11e6-84fc-5890080cc154.png)
![capture d ecran_2016-09-21_16-45-05](https://cloud.githubusercontent.com/assets/15774019/18715990/70bf08f8-801b-11e6-9c7b-b228dbac977c.png)

**Comment by fredbcode** [comment]
Hum, in your previous trace with chrome, it was directly connected to squid, no ?
As far I can tell there is no NTLM exchange, so maybe you are right, but without E2, there is a problem after the 407 requests or maybe I missed something ?

Steep by steep, can you confirm your post "Ce site est inacessible" was without E2 ? If yes do you have the same result with FF ? Do you have a 200/304 code (and the login) in squid's log ? 
With wireshark there are some NTLM exhanges ?

**Comment by ai99** [comment]
I actually made all traces with e2guardian listening on eth2:3128 forwarding to squid on 127.0.0.1:8080.

- When using squid directly there is no problem since all websites are allowed (e2 is the one who denies access to sites). Authentication is correct for http and httpS sites, including YTB where I can look at videos etc.

- When using E2 with no denied sites (bannedsitelist and bannedurllist are empty) I can also surf on all http and httpS sites. Authentication is correctly reported for all sites, including YTB, in both squid and E2's log files like this : 
`2016-09-21T17:07:16.132946+02:00 amon.etb1.lan e2guardian0[2080]: "2016.9.21 17:07:16","admin","10.1.2.50","https://www.youtube.com:443","","CONNECT","99078","0","","5","200","-","","","","","-","-",""`
`2016-09-21T17:07:16.149170+02:00 amon.etb1.lan squid3: 1474470436.149  12136 10.1.2.50 TCP_MISS/200 5871 CONNECT s.youtube.com:443 dompedago\\admin HIER_DIRECT/74.125.133.113 -`


- When I add YTB in bannedsitelist for filter1 and my user is also in filter1 I get a correct E2 denied page with all informations (username, ip, url, etc.)

- When I add YTB in bannedsitelist for filter1 BUT my user is in filter2 I get a truncated E2 denied page while I should not be denied.

This only happens with NTLM auth, when using LDAP auth with squid, everything works fine.

So in my opinion, the problem is related to E2 and the way it manages NTLM auth. Because it does not wait the NTLM auth process to finish before determining the username and switch to the correct filter.

**Comment by fredbcode** [comment]
Ok it's more clear, thanks
Can I see your full configuration somewhere ?
Have you tried with another next group like filter3 or 4 ( with YTB in bannedsitelist in filter1) ?

Unfortunately I'm using Digest/Basic now, perhaps you have a VM snapshot ?

**Comment by ai99** [comment]
Actually "admin" is in filter5, there are 8 filters.

Here is the E2 configuration
[guardian_conf.tar.gz](https://github.com/e2guardian/e2guardian/files/486746/guardian_conf.tar.gz)

Here is the Squid configuration
[squid3.tar.gz](https://github.com/e2guardian/e2guardian/files/486747/squid3.tar.gz)

**Comment by fredbcode** [comment]
Ok
You said filter2 before, it's a new test ? Same result ?

**Comment by ai99** [comment]
I said filter2 to simplify.

Since the opening of this issue all the tests have been made using the same server and the same configuration, only playing with filter1 and filter5. The only modifications are the Ubuntu updates daily applied on the test server.

The 2 configuration archives come from the test server.

**Comment by fredbcode** [comment]
auth_param ntlm program /usr/lib/squid3/ntlm_smb_lm_auth -a  dompedago/scribe domadmin/horus
auth_param ntlm children 20
auth_param basic children 20
auth_param basic realm Squid proxy-caching web server
auth_param basic credentialsttl 2 hours
auth_param basic casesensitive off

Please, can you make it clear to me why I there is a mix between basic and ntlm, without any basic program ?

authplugin = '/etc/guardian/guardian0/authplugins/proxy-ntlm.conf'
authplugin = '/etc/guardian/guardian0/authplugins/proxy-basic.conf'
authplugin = '/etc/guardian/guardian0/common/ip.conf'

Are you using simultaneous identifications method ? Do you have the same issue without "auth_param basic" and authpluging basic and IP ? 

> So in my opinion, the problem is related to E2 and the way it manages NTLM auth. Because it does not wait the NTLM auth process to finish before determining the username and switch to the correct filter.

Yes maybe, however NTLM is popular and this is the first report problem
And in wireshark there is no answer from browser (to late or not) 

I hope I will find time to make a test with NTLM soon

**Comment by ai99** [comment]
Unfortunately I get the same issue when only using 
`authplugin = '/etc/guardian/guardian0/authplugins/proxy-ntlm.conf'`
(basic and ip auth removed).

**Comment by fredbcode** [comment]
I made some tests, this is not related with youtube no matter the website

I banned numsys.eu, for group1

With *HTTP* mode, the first request contains my login "admin"

2016-12-19T17:30:24.998791+01:00 amon.monreseau.lan e2guardian0[26743]: 
"2016.12.19 
17:30:24","admin","172.16.0.30","http://numsys.eu","","GET","33878","-80","","5","200","text/html","","","","","-","-",""

Now exactly the same with "*HTTPS*"

2016-12-19T17:32:36.906608+01:00 amon.monreseau.lan e2guardian0[26743]: 
"2016.12.19 17:32:36","-","172.16.0.30","https://numsys.eu:443","*Accès 
interdit* Blocked HTTPS site: numsys.eu","CONNECT","0","0","","1","403","

As you can see no user


Now, another *HTTPS* website but allowed for all users (just not 
blacklisted)

2016-12-19T17:37:31.097515+01:00 amon.monreseau.lan e2guardian0[26743]: 
"2016.12.19 
17:37:31","admin","172.16.0.30","https://www.sncf.fr:443","","CONNECT","3845","0","","5","200","-","","","","","-","-",""

As far I can tell the blacklist "group 1" is always applied with *HTTPS* 
website and NTLM , the behavior is the same with IE

**Comment by fredbcode** [comment]
For a complet record

If I denied a website only for group2

2016-12-19T17:55:07.227415+01:00 amon.monreseau.lan e2guardian0[26743]: 
"2016.12.19 
17:55:07","admin","172.16.0.30","https://www.cmb.fr:443","*Accès 
interdit* Blocked HTTPS site: 
cmb.fr","CONNECT","0","0","","5","403","-","","","","","-","-",""

In this case "Admin" is well logged and the website is banned

So the situation is more clear now, in HTTPS/NTLM:

- A website banned for the first group is blocked for all users (no 
matters the group) and there is no name in log (so group 1 ...)
- A website banned for a specific group is blocked only for the right 
group -> this is good in this case

**Comment by fredbcode** [comment]
Fixed in 3.5.0
